src/error.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef SNAP_CONFINE_ERROR_H
+#define SNAP_CONFINE_ERROR_H
+
+/**
+ * This module defines APIs for simple error management.
+ *
+ * Errors are allocated objects that can be returned and passed around from
+ * functions.  Errors carry a formatted message and optionally a scoped error
+ * code. The code is coped with a string "domain" that simply acts as a
+ * namespace for various interacting modules.
+ **/
+
+/**
+ * Opaque error structure.
+ **/
+struct sc_error;
+
+/**
+ * Error domain for errors related to system errno.
+ **/
+#define SC_ERRNO_DOMAIN "errno"
+
+/**
+ * Initialize a new error object.
+ *
+ * The domain is a cookie-like string that allows the caller to distinguish
+ * between "namespaces" of error codes. It should be a static string that is
+ * provided by the caller. Both the domain and the error code can be retrieved
+ * later.
+ *
+ * This function calls die() in case of memory allocation failure.
+ **/
+__attribute__ ((format(printf, 3, 4)))
+struct sc_error *sc_error_init(const char *domain, int code, const char *msgfmt,
+			       ...);
+
+/**
+ * Initialize an errno-based error.
+ *
+ * The error carries a copy of errno and a custom error message as designed by
+ * the caller. See sc_error_init() for a more complete description.
+ *
+ * This function calls die() in case of memory allocation failure.
+ **/
+__attribute__ ((format(printf, 2, 3)))
+struct sc_error *sc_error_init_from_errno(int errno_copy, const char *msgfmt,
+					  ...);
+
+/**
+ * Get the error domain out of an error object.
+ *
+ * The error domain acts as a namespace for error codes. The return value must
```

Comment:
nit: I would propose to rephrase the doc string a little bit: "No change of ownership takes place".